File: backend/app/schemas/missions_design.py
"""Design mission schemas (P7-5) — `/missions/design/*`.

One `GET .../attempts/{attempt_id}` returns everything the nine-step UI
needs in a single fetch (design fields, components with their budget
overrides, CONOPS matrix, link entry, and the computed dashboard) — a
deliberate simplification of Madar's one-fetch-per-page shape, since this
is a single-page wizard, not eleven server-rendered HTML files.
"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


# ── component library (the picker) ──────────────────────────────────────

class DesignLibraryComponentOut(BaseModel):
    id: UUID
    component_name: str
    subsystem: str
    tag: str | None = None
    example_role: str | None = None
    scaled_description: str | None = None
    length_mm: float | None = None
    width_mm: float | None = None
    height_mm: float | None = None
    scaled_mass_g: float | None = None
    voltage_v: float | None = None
    current_ma: float | None = None
    data_size: str | None = None
    assumed_cost_usd: float | None = None
    temperature_range: str | None = None
    key_specs: str | None = None
    image_url: str | None = None
    component_code: str | None = None
    datasheet_url: str | None = None


# ── design components (with budget overrides) ────────────────────────────

class DesignDataEntryOut(BaseModel):
    data_type: str | None = None
    data_size_per_measurement_kb: float = 0.0
    measurements_per_minute: float = 0.0
    priority: str = "Medium"
    storage_mode: Literal["Stored", "Sent", "Both"] = "Stored"
    notes: str | None = None


class DesignPowerEntryOut(BaseModel):
    voltage_v: float | None = None
    current_ma: float | None = None
    notes: str | None = None


class DesignMassEntryOut(BaseModel):
    quantity: int | None = None
    mass_per_unit_g: float | None = None
    length_mm: float | None = None
    width_mm: float | None = None
    height_mm: float | None = None
    notes: str | None = None


class DesignCostEntryOut(BaseModel):
    quantity: int | None = None
    cost_per_unit_aed: float | None = None
    vendor: str | None = None
    priority: str | None = None
    purchase_link: str | None = None
    notes: str | None = None


class DesignComponentOut(BaseModel):
    id: UUID
    library_component_id: UUID
    component_name: str
    subsystem: str
    image_url: str | None = None
    quantity: int
    mass_per_unit_g: float | None = None
    length_mm: float | None = None
    width_mm: float | None = None
    height_mm: float | None = None
    voltage_v: float | None = None
    current_ma: float | None = None
    cost_per_unit_aed: float | None = None
    on_mode_ids: list[UUID] = []
    data_entry: DesignDataEntryOut | None = None
    power_entry: DesignPowerEntryOut | None = None
    mass_entry: DesignMassEntryOut | None = None
    cost_entry: DesignCostEntryOut | None = None


class DesignComponentAddIn(BaseModel):
    library_component_id: UUID
    quantity: int = 1


# ── CONOPS ────────────────────────────────────────────────────────────────

class DesignModeOut(BaseModel):
    id: UUID
    mode_name: str
    position: int
    duration_min: float
    description: str | None = None


class ConopsSaveIn(BaseModel):
    mode_durations: dict[UUID, float]  # mode_id -> duration_min
    cell_states: dict[UUID, dict[UUID, bool]]  # component_id -> {mode_id: is_on}


# ── budget entry saves ────────────────────────────────────────────────────

class DataEntrySaveIn(BaseModel):
    data_type: str | None = None
    data_size_per_measurement_kb: float = 0.0
    measurements_per_minute: float = 0.0
    priority: str = "Medium"
    storage_mode: Literal["Stored", "Sent", "Both"] = "Stored"
    notes: str | None = None


class PowerEntrySaveIn(BaseModel):
    voltage_v: float | None = None
    current_ma: float | None = None
    notes: str | None = None


class MassEntrySaveIn(BaseModel):
    quantity: int | None = None
    mass_per_unit_g: float | None = None
    length_mm: float | None = None
    width_mm: float | None = None
    height_mm: float | None = None
    notes: str | None = None


class CostEntrySaveIn(BaseModel):
    quantity: int | None = None
    cost_per_unit_aed: float | None = None
    vendor: str | None = None
    priority: str | None = None
    purchase_link: str | None = None
    notes: str | None = None


class LinkEntrySaveIn(BaseModel):
    band_profile: str = "UHF"
    downlink_frequency_mhz: float
    uplink_frequency_mhz: float
    satellite_antenna_gain_dbi: float
    data_rate_kbps: float
    required_signal_quality_db: float
    notes: str | None = None


class LinkEntryOut(BaseModel):
    band_profile: str
    downlink_frequency_mhz: float
    uplink_frequency_mhz: float
    satellite_antenna_gain_dbi: float
    data_rate_kbps: float
    required_signal_quality_db: float
    notes: str | None = None
    is_saved: bool


# ── design fields ──────────────────────────────────────────────────────

class DesignUpdateIn(BaseModel):
    design_name: str | None = None
    design_objective: str | None = None
    orbit_type: str | None = None
    orbit_duration_min: float | None = None
    orbits_per_day: float | None = None
    selected_cubesat_size: Literal["1U", "2U", "3U", "6U"] | None = None
    selected_solar_cells: int | None = None
    battery_capacity_wh: float | None = None


# ── the dashboard (composed summary) ────────────────────────────────────

class StepStatusOut(BaseModel):
    has_data: bool
    is_valid: bool


class DataBudgetSummaryOut(BaseModel):
    total_per_orbit_kb: float
    total_per_day_kb: float
    total_stored_per_day_kb: float
    total_sent_per_day_kb: float
    storage_remaining_kb: float
    max_storage_kb: float
    required_storage_margin_kb: float


class PowerBudgetSummaryOut(BaseModel):
    total_power_mw: float
    total_energy_per_orbit_mwh: float
    total_energy_per_day_mwh: float
    power_margin_mw: float
    required_solar_cells: int
    generated_power_mw: float
    selected_solar_cells: int
    power_per_solar_cell_w: float


class MassBudgetSummaryOut(BaseModel):
    total_mass_kg: float
    mass_margin_kg: float
    total_volume_cm3: float
    volume_margin_cm3: float
    max_allowed_mass_kg: float
    available_internal_volume_cm3: float


class CostBudgetSummaryOut(BaseModel):
    total_cost_aed: float
    cost_margin_aed: float
    maximum_budget_aed: float


class LinkBudgetSummaryOut(BaseModel):
    margin_db: float
    status: str
    # The bar, not just the score. Showing "Margin: 0.61 dB - Failed Link"
    # without saying 3 dB is required leaves a student guessing which way to
    # move, which is the opposite of a teaching tool.
    good_threshold_db: float
    weak_threshold_db: float
    assumed_distance_km: float
    transmit_power_dbm: float


class EnergyBudgetSummaryOut(BaseModel):
    sunlit_minutes: float
    eclipse_minutes: float
    generated_per_orbit_mwh: float
    consumed_per_orbit_mwh: float
    energy_margin_mwh: float
    energy_balance_ok: bool
    eclipse_draw_mwh: float
    battery_capacity_mwh: float
    depth_of_discharge_pct: float
    max_depth_of_discharge_pct: float
    depth_of_discharge_ok: bool


class DownlinkBudgetSummaryOut(BaseModel):
    data_to_downlink_per_orbit_kb: float
    downlink_capacity_per_orbit_kb: float
    downlink_margin_kb: float
    contact_minutes: float
    utilisation_pct: float


class ConopsSummaryOut(BaseModel):
    total_mode_duration_min: float
    duration_difference_min: float


class MarginRowOut(BaseModel):
    key: str
    label: str
    value: float
    unit: str
    status: str  # good | tight | fail | incomplete
    interpretation: str


class ModuleCardOut(BaseModel):
    key: str
    title: str
    status: str
    kpi1_label: str
    kpi1_value: str
    kpi2_label: str
    kpi2_value: str
    tab: str


class AlertOut(BaseModel):
    severity: str  # error | warning | info | success
    step: str | None = None
    message: str


class RecommendationOut(BaseModel):
    key: str
    title: str
    message: str
    why: str


class OverallStatusOut(BaseModel):
    label: str
    all_valid: bool
    errors: int
    warnings: int
    incomplete: int


class DashboardOut(BaseModel):
    all_valid: bool
    steps: dict[str, StepStatusOut]
    conops: ConopsSummaryOut
    data: DataBudgetSummaryOut
    power: PowerBudgetSummaryOut
    mass: MassBudgetSummaryOut
    cost: CostBudgetSummaryOut
    link: LinkBudgetSummaryOut
    energy: EnergyBudgetSummaryOut
    downlink: DownlinkBudgetSummaryOut

    # Design v2 (7D-3) — the payoff screen. Madar had all of this and the
    # port rendered a boolean; the numbers were already here, the judgement
    # was not.
    overall: OverallStatusOut
    kpis: dict
    margins: list[MarginRowOut] = []
    module_cards: list[ModuleCardOut] = []
    charts: dict = {}
    alerts: list[AlertOut] = []
    recommendations: list[RecommendationOut] = []


# ── the full state fetch ─────────────────────────────────────────────────

class CubeSatPresetOut(BaseModel):
    size: str
    available_volume_cm3: float
    max_mass_kg: float


class DesignStateOut(BaseModel):
    id: UUID
    attempt_id: UUID
    mission_id: UUID
    variant_id: UUID
    variant_label: str
    attempt_status: str
    design_name: str
    design_objective: str | None = None
    orbit_type: str | None = None
    orbit_duration_min: float | None = None
    orbits_per_day: float | None = None
    selected_cubesat_size: str
    selected_solar_cells: int
    battery_capacity_wh: float | None = None
    created_at: datetime | None = None

    components: list[DesignComponentOut] = []
    modes: list[DesignModeOut] = []
    link_entry: LinkEntryOut | None = None

    cubesat_presets: list[CubeSatPresetOut] = []
    band_presets: dict[str, dict] = {}

    dashboard: DashboardOut
    # F9 — what the model simplifies, said out loud rather than hidden.
    assumptions: list[str] = []
    # 2026-08-17 — step_key -> is_unlocked, all 9 DESIGN_STEP_LABELS keys.
    # An attempt outside any cohort (or a cohort with no gates set) reads
    # all True — gating is opt-in per cohort, not a default-locked system.
    # Drives DesignMissionPage.tsx's tab-blocking alongside the existing
    # data-dependency check.
    step_gates: dict[str, bool] = {}


# ── Teaching surfaces (Design v2, 7D-4 / 7D-5) ──────────────────────────

class DesignBriefingOut(BaseModel):
    """Read before designing. Served without an attempt existing, so opening
    it never burns a retry."""

    mission_id: UUID
    mission_title: str
    mission_summary: str | None = None
    variant_id: UUID
    variant_label: str
    points: int
    what_is_a_budget: str
    step_order: list[dict]
    limits: list[dict]
    cubesat_sizes: list[dict]
    budgets: list[dict]
    assumptions: list[str]


class DesignHandbookOut(BaseModel):
    """The Design Handbook — open while designing, disclosure scaled by
    variant exactly as the operate mission's Ops Handbook is."""

    disclosure: str
    what_is_a_budget: str
    step_order: list[dict]
    budgets: list[dict]
    data_types: list[dict]
    mistakes: list[dict]
    assumptions: list[str]


# ── Component library administration (Design v2, 7D-7) ──────────────────

class LibraryComponentBase(BaseModel):
    component_name: str | None = None
    subsystem: str | None = None
    tag: str | None = None
    example_role: str | None = None
    scaled_description: str | None = None
    length_mm: float | None = None
    width_mm: float | None = None
    height_mm: float | None = None
    scaled_mass_g: float | None = None
    voltage_v: float | None = None
    current_ma: float | None = None
    data_size: str | None = None
    assumed_cost_usd: float | None = None
    temperature_range: str | None = None
    key_specs: str | None = None
    component_code: str | None = None
    datasheet_url: str | None = None
    notes: str | None = None


class LibraryComponentCreateIn(LibraryComponentBase):
    component_name: str
    subsystem: str


class LibraryComponentUpdateIn(LibraryComponentBase):
    is_active: bool | None = None


class LibraryComponentAdminOut(LibraryComponentBase):
    id: UUID
    component_name: str
    subsystem: str
    is_active: bool
    image_url: str | None = None
    updated_at: datetime | None = None
    updated_by_name: str | None = None
    # How many designs have ever added this. Retiring is safe at any count;
    # this is here so an editor can see the blast radius of a *spec change*
    # before making one.
    used_in_designs: int = 0


class LibraryBulkImportIn(BaseModel):
    components: list[LibraryComponentCreateIn]


class LibraryBulkImportOut(BaseModel):
    created: int
    updated: int
    errors: list[str] = []

File: frontend/src/api/missionsDesign.ts
/** Design mission API (P7-5) — thin wrappers over `/missions/design/*`.
 * Types mirror `schemas/missions_design.py` field for field.
 */
import { api } from "./client";

export interface DesignLibraryComponent {
  id: string;
  component_name: string;
  subsystem: string;
  tag: string | null;
  example_role: string | null;
  scaled_description: string | null;
  length_mm: number | null;
  width_mm: number | null;
  height_mm: number | null;
  scaled_mass_g: number | null;
  voltage_v: number | null;
  current_ma: number | null;
  data_size: string | null;
  assumed_cost_usd: number | null;
  temperature_range: string | null;
  key_specs: string | null;
  image_url: string | null;
  component_code: string | null;
  datasheet_url: string | null;
}

export interface DesignDataEntry {
  data_type: string | null;
  data_size_per_measurement_kb: number;
  measurements_per_minute: number;
  priority: string;
  storage_mode: "Stored" | "Sent" | "Both";
  notes: string | null;
}

export interface DesignPowerEntry {
  voltage_v: number | null;
  current_ma: number | null;
  notes: string | null;
}

export interface DesignMassEntry {
  quantity: number | null;
  mass_per_unit_g: number | null;
  length_mm: number | null;
  width_mm: number | null;
  height_mm: number | null;
  notes: string | null;
}

export interface DesignCostEntry {
  quantity: number | null;
  cost_per_unit_aed: number | null;
  vendor: string | null;
  priority: string | null;
  purchase_link: string | null;
  notes: string | null;
}

export interface DesignComponent {
  id: string;
  library_component_id: string;
  component_name: string;
  subsystem: string;
  image_url: string | null;
  quantity: number;
  mass_per_unit_g: number | null;
  length_mm: number | null;
  width_mm: number | null;
  height_mm: number | null;
  voltage_v: number | null;
  current_ma: number | null;
  cost_per_unit_aed: number | null;
  on_mode_ids: string[];
  data_entry: DesignDataEntry | null;
  power_entry: DesignPowerEntry | null;
  mass_entry: DesignMassEntry | null;
  cost_entry: DesignCostEntry | null;
}

export interface DesignMode {
  id: string;
  mode_name: string;
  position: number;
  duration_min: number;
  description: string | null;
}

export interface LinkEntry {
  band_profile: string;
  downlink_frequency_mhz: number;
  uplink_frequency_mhz: number;
  satellite_antenna_gain_dbi: number;
  data_rate_kbps: number;
  required_signal_quality_db: number;
  notes: string | null;
  is_saved: boolean;
}

export interface CubeSatPreset {
  size: string;
  available_volume_cm3: number;
  max_mass_kg: number;
}

export interface BandPreset {
  downlink_frequency_mhz: number;
  uplink_frequency_mhz: number;
  satellite_antenna_gain_dbi: number;
  data_rate_kbps: number;
  required_signal_quality_db: number;
}

export interface StepStatus {
  has_data: boolean;
  is_valid: boolean;
}

export interface Dashboard {
  all_valid: boolean;
  steps: Record<string, StepStatus>;
  conops: { total_mode_duration_min: number; duration_difference_min: number };
  data: {
    total_per_orbit_kb: number; total_per_day_kb: number; total_stored_per_day_kb: number;
    total_sent_per_day_kb: number; storage_remaining_kb: number; max_storage_kb: number;
    required_storage_margin_kb: number;
  };
  power: {
    total_power_mw: number; total_energy_per_orbit_mwh: number; total_energy_per_day_mwh: number;
    power_margin_mw: number; required_solar_cells: number; generated_power_mw: number;
    selected_solar_cells: number; power_per_solar_cell_w: number;
  };
  mass: {
    total_mass_kg: number; mass_margin_kg: number; total_volume_cm3: number;
    volume_margin_cm3: number; max_allowed_mass_kg: number; available_internal_volume_cm3: number;
  };
  cost: { total_cost_aed: number; cost_margin_aed: number; maximum_budget_aed: number };
  link: {
    margin_db: number; status: string;
    good_threshold_db: number; weak_threshold_db: number;
    assumed_distance_km: number; transmit_power_dbm: number;
  };
  /** Design v2 (7D-2, F8/D4) — energy over a whole orbit, and the battery. */
  energy: {
    sunlit_minutes: number; eclipse_minutes: number;
    generated_per_orbit_mwh: number; consumed_per_orbit_mwh: number;
    energy_margin_mwh: number; energy_balance_ok: boolean;
    eclipse_draw_mwh: number; battery_capacity_mwh: number;
    depth_of_discharge_pct: number; max_depth_of_discharge_pct: number; depth_of_discharge_ok: boolean;
  };
  /** Design v2 (7D-2, F7) — can you actually get the data down? */
  downlink: {
    data_to_downlink_per_orbit_kb: number; downlink_capacity_per_orbit_kb: number;
    downlink_margin_kb: number; contact_minutes: number; utilisation_pct: number;
  };
  /** Design v2 (7D-3) — the payoff screen Madar had and the port dropped. */
  overall: { label: string; all_valid: boolean; errors: number; warnings: number; incomplete: number };
  kpis: Record<string, number>;
  margins: MarginRow[];
  module_cards: ModuleCard[];
  charts: DesignCharts;
  alerts: DesignAlert[];
  recommendations: Recommendation[];
}

export type MarginStatus = "good" | "tight" | "fail" | "incomplete";

export interface MarginRow {
  key: string;
  label: string;
  value: number;
  unit: string;
  status: MarginStatus;
  interpretation: string;
}

export interface ModuleCard {
  key: string;
  title: string;
  status: MarginStatus;
  kpi1_label: string;
  kpi1_value: string;
  kpi2_label: string;
  kpi2_value: string;
  tab: string;
}

export interface DesignAlert {
  severity: "error" | "warning" | "info" | "success";
  step: string | null;
  message: string;
}

export interface Recommendation {
  key: string;
  title: string;
  message: string;
  why: string;
}

export interface DesignCharts {
  power_by_subsystem: { subsystem: string; value: number }[];
  mass_by_subsystem: { subsystem: string; value: number }[];
  cost_by_subsystem: { subsystem: string; value: number }[];
  mode_distribution: { mode_name: string; duration_min: number }[];
}

export interface DesignState {
  id: string;
  attempt_id: string;
  mission_id: string;
  variant_id: string;
  variant_label: string;
  attempt_status: "in_progress" | "submitted" | "passed" | "failed" | "abandoned";
  design_name: string;
  design_objective: string | null;
  orbit_type: string | null;
  orbit_duration_min: number | null;
  orbits_per_day: number | null;
  selected_cubesat_size: string;
  selected_solar_cells: number;
  battery_capacity_wh: number | null;
  created_at: string | null;
  components: DesignComponent[];
  modes: DesignMode[];
  link_entry: LinkEntry | null;
  cubesat_presets: CubeSatPreset[];
  band_presets: Record<string, BandPreset>;
  dashboard: Dashboard;
  /** F9 — what this teaching model simplifies, stated rather than hidden. */
  assumptions: string[];
  /** 2026-08-17 — step_key -> is_unlocked, all 9 steps. An attempt outside
   * any cohort (or a cohort with no gates set) reads all true. */
  step_gates: Record<string, boolean>;
}

export async function fetchDesignState(attemptId: string): Promise<DesignState> {
  const { data } = await api.get<DesignState>(`/missions/design/attempts/${attemptId}`);
  return data;
}

export async function fetchDesignLibrary(params?: { subsystem?: string; search?: string }): Promise<DesignLibraryComponent[]> {
  const { data } = await api.get<DesignLibraryComponent[]>("/missions/design/library", { params });
  return data;
}

export async function updateDesign(attemptId: string, body: Partial<{
  design_name: string; design_objective: string; orbit_type: string;
  orbit_duration_min: number; orbits_per_day: number;
  selected_cubesat_size: string; selected_solar_cells: number;
  battery_capacity_wh: number;
}>): Promise<DesignState> {
  const { data } = await api.patch<DesignState>(`/missions/design/attempts/${attemptId}`, body);
  return data;
}

export async function addDesignComponent(attemptId: string, libraryComponentId: string, quantity = 1): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/components`, {
    library_component_id: libraryComponentId, quantity,
  });
  return data;
}

export async function removeDesignComponent(attemptId: string, designComponentId: string): Promise<DesignState> {
  const { data } = await api.delete<DesignState>(`/missions/design/attempts/${attemptId}/components/${designComponentId}`);
  return data;
}

export async function saveConops(
  attemptId: string, modeDurations: Record<string, number>, cellStates: Record<string, Record<string, boolean>>,
): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/conops`, {
    mode_durations: modeDurations, cell_states: cellStates,
  });
  return data;
}

export async function saveDataBudget(attemptId: string, componentId: string, body: Partial<DesignDataEntry>): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/components/${componentId}/data-budget`, body);
  return data;
}

export async function savePowerBudget(attemptId: string, componentId: string, body: Partial<DesignPowerEntry>): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/components/${componentId}/power-budget`, body);
  return data;
}

export async function saveMassBudget(attemptId: string, componentId: string, body: Partial<DesignMassEntry>): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/components/${componentId}/mass-budget`, body);
  return data;
}

export async function saveCostBudget(attemptId: string, componentId: string, body: Partial<DesignCostEntry>): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/components/${componentId}/cost-budget`, body);
  return data;
}

export async function saveLinkBudget(attemptId: string, body: Omit<LinkEntry, "is_saved">): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/link-budget`, body);
  return data;
}

export async function completeDesign(attemptId: string): Promise<DesignState> {
  const { data } = await api.post<DesignState>(`/missions/design/attempts/${attemptId}/complete`);
  return data;
}

// ── Teaching surfaces (Design v2, 7D-4 / 7D-5) ──────────────────────────

export interface DesignBriefing {
  mission_id: string;
  mission_title: string;
  mission_summary: string | null;
  variant_id: string;
  variant_label: string;
  points: number;
  what_is_a_budget: string;
  step_order: { key: string; label: string; detail: string; depends_on: string[] }[];
  limits: { key: string; label: string; value: string; detail: string }[];
  cubesat_sizes: { size: string; max_mass_kg: number; available_volume_cm3: number }[];
  budgets: { key: string; title: string; checks: string; why_it_matters: string }[];
  assumptions: string[];
}

export interface HandbookBudget {
  key: string;
  title: string;
  checks: string;
  formula: string;
  means?: string;
  fails_when?: string;
  why_it_matters?: string;
  fix?: string;
}

export interface HandbookMistake {
  key: string;
  title: string;
  symptom: string;
  steps: string[];
  meaning?: string;
  fix?: string;
}

export interface DesignHandbook {
  disclosure: "full" | "symptoms" | "reference";
  what_is_a_budget: string;
  step_order: { key: string; label: string; detail: string; depends_on: string[] }[];
  budgets: HandbookBudget[];
  data_types: { name: string; detail: string }[];
  mistakes: HandbookMistake[];
  assumptions: string[];
}

export async function fetchDesignBriefing(missionId: string, variantId?: string): Promise<DesignBriefing> {
  const { data } = await api.get<DesignBriefing>(`/missions/design/briefing/${missionId}`, {
    params: variantId ? { variant_id: variantId } : undefined,
  });
  return data;
}

export async function fetchDesignHandbook(attemptId: string): Promise<DesignHandbook> {
  const { data } = await api.get<DesignHandbook>(`/missions/design/attempts/${attemptId}/handbook`);
  return data;
}
